### PR TITLE
[Addon] pyroscope addon

### PR DESCRIPTION
**Pull request by fourierr**
Signed-off-by: Xiangbo Ma <maxiangboo@cmbchina.com>

### Description of your changes
 pyroscope addon

### Checklist

I have:

- [x] Title of the PR starts with type (e.g. `[Addon]` or `[Trait]`).
- [ ] Updated/Added any relevant [documentation] and [examples].
- [x] Unit/E2E Tests added.

### CODE REVIEW

**Comment by wonderflow** [comment]
Great work! Ping @wangyikewxgm

**Comment by oeular** [comment]
I think pyroscope is just as a trait  to workload, so it should not be deployed on the helm, because helm controller is optional.
Here is the way  I think it should be.

apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: app-show
  namespace: real-new-ns
spec:
  components:
    - name: web-show
      type: webservice
      properties:
        exposeType: ClusterIP
        image: beellzrocks/shippingservice
        env:
        - name: PORT
          value: "50051"
        - name: APPLICATION_NAME # Application name shown in the Pyroscope UI
          value: "web-show"
        - name: SERVER_ADDRESS # To change Pyroscope Server Port change the value, the naming rule is pyroscope-<component name>
          value: "http://pyroscope-web-show:4040"
      traits:
        - type: pyroscope

**Comment by fourierr** [comment]
> Application name shown in the Pyroscope UI

- The env is no question, what about pyroscope trait do?
- Just only installing pyroscope in your cluster, every app can use it instead of declare a trait.
- Yeas, when you want to deliver some param to pyroscope , you neet to use a trait. But this config param is for `pyroscope sever` otherwise `pyroscope client`. And the client is configured by env according to your description.
-  So i think trait for pyroscope is  unnecessary. It's like that wo don't need declare an ocm trait, but we can use ocm when you enable it.  ping @oeular

**Comment by oeular** [comment]
In vela, the capacity needed by workload(such as webservice, worker, task, cloneset) is more suitable as a trait.
For the all components of vela itself, pyroscope can be installed by helm as an applicaion

**Comment by oeular** [comment]
If a trait for workload, here I think it should be: https://github.com/kubevela/catalog/pull/363#issue-1260952527

**Comment by fourierr** [comment]
我看了一下你的方式，如果按照这样把pyroscoep做成一个trait意味着每个组件都会创建一套完整pyroscope server包括deployemnt、serveice、pv/pvc，我觉得这个成本是不可接受的。
我们为什么不让多个应用共用一个pyroscope server呢？就像是prometheus，一个集群中只有一个prometheus server，而不是每个应用都有独自的prometheus server。
@oeular

File: test/e2e-test/hack/addon-vela-test.sh [hunk 1]
```diff
@@ -49,6 +49,22 @@ vela addon enable cert-manager
...
+# enable pyroscope
+kubectl create ns pyroscope-system
+declare -x DEFAULT_VELA_NS=pyroscope-system
+vela addon enable fluxcd
+vela addon enable pyroscope
+declare -x DEFAULT_VELA_NS=vela-system
+
+# disable pyroscope
+declare -x DEFAULT_VELA_NS=pyroscope-system
+vela addon disable pyroscope
```
Comment (wangyikewxgm) [inline test/e2e-test/hack/addon-vela-test.sh:63]: Is this addon should be installed `pyroscope-system` namespace

File: addons/pyroscope/readme.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+```shell
+vela addon disable pyroscope
+```
+
+## Note
+### Pyroscope for Golang applications
+
+- To start profiling a Go application, you need to include our go module in your app
+    ```shell
```
Comment (wangyikewxgm) [inline addons/pyroscope/readme.md:21]: What's the vela application using this trait look like? Can you put a demo in example? I truly suggest add a demo in example/ dir

**Comment by wangyikewxgm** [comment]
This pr doesn't contain the tratDefintion?

File: test/e2e-test/hack/addon-vela-test.sh (hunk 1 above)
Comment (fourierr) [reply test/e2e-test/hack/addon-vela-test.sh:63]: I refer to fluxcd-system，and a separate ns is more clear to user？

**Comment by fourierr** [comment]
> This pr doesn't contain the tratDefintion?

Yes, after discussion with @wonderflow , it is not need a tratDefintion. 
It seems to like ocm, when you enable it in rhe cluster, any app can use the pyroscope via modify your code,

File: addons/pyroscope/readme.md (hunk 2 above)
Comment (fourierr) [reply addons/pyroscope/readme.md:21]: As described by https://github.com/kubevela/catalog/pull/361#issuecomment-1147019554

File: test/e2e-test/hack/addon-vela-test.sh (hunk 1 above)
Comment (fourierr) [reply test/e2e-test/hack/addon-vela-test.sh:63]: It is default to `vela-system`

File: addons/pyroscope/metadata.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+description: Pyroscope is an open source platform, consisting of server and agent. It allows the user to collect, store, and query the profiling data in a CPU and disk efficient way.
+url: https://github.com/pyroscope-io/pyroscope
+icon: https://raw.githubusercontent.com/pyroscope-io/pyroscope/main/webapp/images/favicon.ico
+
+tags:
+  - profiling tool
+
+deployTo:
+  control_plane: true
+
```
Comment (wangyikewxgm) [inline addons/pyroscope/metadata.yaml:12]: Need install to runtime-cluster？

File: addons/pyroscope/readme.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  pyroscope.Start(pyroscope.Config{
+    ApplicationName: "simple.golang.app",
+    // replace this with the address of pyroscope server
+    ServerAddress:   "http://pyroscope-server:4040",
+    // you can disable logging by setting this to nil
+    Logger:          pyroscope.StandardLogger,
+
+    // optionally, if authentication is enabled, specify the API key:
+    // AuthToken: os.Getenv("PYROSCOPE_AUTH_TOKEN"),
+
```
Comment (wangyikewxgm) [inline addons/pyroscope/readme.md:41]: I think developer can read these configuration from env. Then we can support a trait patch these env to a container.
Then app use this trait by
```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: first-vela-app
spec:
  components:
    - name: express-server
      type: webservice
      properties:
        image: oamdev/hello-world
        ports:
         - port: 8000
           expose: true
      traits:
        - type: scaler
          properties:
            replicas: 1
        - type: pyroscope
           server: "http://pyroscope-server:4040"
           appName: context.name
           logger: "pyroscope.StandardLogger"
           ....
```

Comment (fourierr) [reply addons/pyroscope/readme.md:41]: Yes, it's a convenient way for user. I modify it. @wangyikewxgm

File: addons/pyroscope/readme.md [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+        image: nginx:latest
+        ports:
+          - expose: true
+            port: 80
+            protocol: TCP
+        imagePullPolicy: IfNotPresent
+      traits:
+        - type: pyroscope
+          properties:
+            server: "http://pyroscope-server:9084"
```
Comment (wonderflow) [inline addons/pyroscope/readme.md:46]: does every cluster need such a sever when we specify the trait? How it works in multi-cluster delivery case？

Comment (fourierr) [reply addons/pyroscope/readme.md:46]: Multi-cluster client has the ability to push data to one server, user should manually publish domian and create ingress.

File: test/e2e-test/hack/addon-vela-test.sh [hunk 6]
```diff
@@ -49,6 +49,22 @@ vela addon enable cert-manager
 vela addon enable flink-kubernetes-operator
 # set back to the DEFAULT_VELA_NS
 declare -x DEFAULT_VELA_NS=vela-system
+
+# test pyroscope addon
+# enable pyroscope
+kubectl create ns pyroscope-system
+declare -x DEFAULT_VELA_NS=pyroscope-system
+vela addon enable fluxcd
+vela addon enable pyroscope
```
Comment (wangyikewxgm) [inline test/e2e-test/hack/addon-vela-test.sh:58]: As discussed with @fourierr. These disable/enable fluxcd addon will waste too much time for running CI. The next pr will refactor the test script to reduce these operation.

Comment (fourierr) [reply test/e2e-test/hack/addon-vela-test.sh]: > As discussed with @fourierr. These disable/enable fluxcd addon will waste too much time for running CI. The next pr will refactor the test script to reduce these operation.

Yes, i will adjust the order of CI in next pr.

**Review by wangyikewxgm — APPROVED** [review]
LGTM, great job.

**Pull request merged**